crates/core/tedge_api/src/workflow/script.rs

```diff
@@ -265,6 +246,41 @@ impl ExitHandlers {
     }
 }
 
+/// Extract the json output of a script outcome
+pub fn extract_json_output(
+    program: &str,
+    outcome: std::io::Result<std::process::Output>,
+) -> Result<Value, String> {
+    match outcome {
+        Ok(output) => match output.status.code() {
+            None => Err(format!(
+                "Program {program} has been killed by SIG{}",
+                output.status.signal().unwrap_or(0) as u8
+            )),
+            Some(0) => json_stdout_excerpt(program, output.stdout),
+            Some(code) => Err(format!("Program {program} failed with exit code {code}")),
+        },
+        Err(err) => Err(format!("Program {program} cannot be launched: {err}")),
+    }
+}
+
+fn json_stdout_excerpt(program: &str, stdout: Vec<u8>) -> Result<Value, String> {
+    match String::from_utf8(stdout) {
+        Err(_) => Err(format!("{program} returned no UTF8 stdout")),
+        Ok(content) => match extract_script_output(content) {
+            None => Err(format!(
+                "{program} returned no :::tedge::: content on stdout"
+            )),
+            Some(excerpt) => match serde_json::from_str(&excerpt) {
+                Ok(json) => Ok(json),
+                Err(err) => Err(format!(
+                    "{program} returned non JSON content on stdout: {err}"
+                )),
+            },
+        },
+    }
+}
```

Comment:
suggestion: Here, `program` is only passed to append it to the error message. Instead, as the error is opaque, we could easily use `anyhow::Error` and attach `.context()` with program name in calling code.

Also, I think nesting matches makes it a bit hard to read due to Ok and Err arms being in arbitrary order; would you agree that it's more readable when written as series of steps with early return?

```rust
/// Extract the json output of a script outcome
pub fn extract_json_output(
    outcome: std::io::Result<std::process::Output>,
) -> anyhow::Result<Value> {
    let output = outcome.context("Program cannot be launched")?;

    let exit_code = output
        .status
        .code()
        .with_context(|| format!("Killed by SIG{}", output.status.signal().unwrap_or(0) as u8))?;

    anyhow::ensure!(exit_code != 0, "Failed with exit code {exit_code}");

    let stdout = String::from_utf8(output.stdout).context("Stdout not UTF8")?;

    let excerpt = extract_script_output(stdout)
        .context("Program returned no :::tedge::: content on stdout")?;

    let json =
        serde_json::from_str(&excerpt).context("Program returned non JSON content on stdout")?;

    Ok(json)
}
```

And program name can be attached as context by the caller:

```rust
match extract_json_output(&script.command, output).with_context(|| format!("Could not extract output of program {}", &script.command)) {
    Ok(init_state) => init_state,
    Err(reason) => {
        let err_state = state.update(GenericStateUpdate::failed(reason));
        self.publish_command_state(err_state).await?;
        return Ok(());
    }
}
```

Of course, one would first have to add `anyhow` here, and this probably goes into `reason` field, which is a single string, so we'd have to loop through `source()` of `anyhow::Error` appending it to a single string, so maybe it's not workable. But still, I'd argue removing needless matches makes it more readable.

Reply:
> Also, I think nesting matches makes it a bit hard to read due to Ok and Err arms being in arbitrary order; would you agree that it's more readable when written as series of steps with early return?

This is a matter of taste. With early returns is easy to miss one of these returns especially when hidden being a `?`. However, I tried to follow your suggestion.

See: https://github.com/thin-edge/thin-edge.io/pull/2792/commits/10affc17b54d784c91c6caea6b6a2abae97b6484

> suggestion: Here, `program` is only passed to append it to the error message. Instead, as the error is opaque, we could easily use `anyhow::Error` and attach `.context()` with program name in calling code. Of course, one would first have to add `anyhow` here, and this probably goes into `reason` field, which is a single string, so we'd have to loop through `source()` of `anyhow::Error` appending it to a single string, so maybe it's not workable. But still, I'd argue removing needless matches makes it more readable.

Indeed the `program` parameter is better handled with a context. Using `anyhow` being not adapted here (for the reason you cited), I introduced a poor-man context specialized on `String`. I also used this context only on private functions.

See: https://github.com/thin-edge/thin-edge.io/pull/2792/commits/10affc17b54d784c91c6caea6b6a2abae97b6484